src/core/generate-reducers.ts

```
@@ -74,19 +109,61 @@ const generateReducersFromDefs = (defs: RdxDefinition[], prefix = ``) => {
   let i = reducers.length
 
   while(i--) {
-    const { name, keys, reducerState, reducerHandlers } = reducers[i]
+    const { name, keys, reducerState, reducerHandlers, isApiReducer } = reducers[i]
+
+    acc[name] = {
+      handlers: reducerHandlers,
+      state: isApiReducer ? apiState : reducerState,
+    }
 
-    acc[name] = { state: reducerState, handlers: reducerHandlers }
     let j = keys.length
 
     while (j--) {
       const { handlers } = keys[j]
 
       Object.assign(acc[name].handlers, handlers)
+    }
+
+    if (isApiReducer) {
+      const apiHandlers = acc[name].handlers
+      const apiHandlerKeys = Object.keys(apiHandlers)
+
+      const successHandlerKey = apiHandlerKeys.find(key => key.endsWith(`SUCCESS`))
+      const requestHandlerKey = apiHandlerKeys.find(key => key.endsWith(`REQUEST`))
+      const failureHandlerKey = apiHandlerKeys.find(key => key.endsWith(`FAILURE`))
+      const resetHandlerKey = apiHandlerKeys.find(key => key.startsWith(`RESET`))
+
+      const notAlreadyAHandler = notAnyOf([
+        successHandlerKey,
+        requestHandlerKey,
+        failureHandlerKey,
+        resetHandlerKey,
+      ])
+
+      const reducedApiHandlers = apiHandlerKeys.reduce((acc, key) => {
+        if (notAlreadyAHandler(key)) {
```

Hmm, if I'm reading this correctly this is for the use case where the dev passes in something that overrides the builtin/default API handlers?

If that's the case then I would just throw an error for this one off case. They're better off knowing that what they're doing is invalid instead of having their handlers mysteriously not work.

---

You're reading the code right, but in this context, this code is only dealing with handlers that are code generated to avoid duplicate handlers that can result from the names of keys which are arbitrary

when using `createApiReducer`, developers are free to overwrite the handlers :)

---

isn't that an error though? You probably never want duplicates to happen like that in an single app.

---

sorry, i misread my own stuff.

these things are found in apiHandlerKeys - i set them there because it's a different iteration context and it would be a real ... neat challenge to get access to what's in there while in this block considering the keys are arbitrary.

i put these api types there intentionally, and this removes them, because for ease of writing, they are bundled in with others that may be there and otherwise treated as the same.

this code should be here, because otherwise, RDX will bundle them as duplicates but with undefined handler functions. developers are totally safe and can add whatever keys - they won't get this api behavior if they create a modified copy of `apiState` because it uses an `Object.is` comparison.